msdk-featdet/msdk-featdet-ADAP-3D/src/main/java/io/github/msdk/featdet/ADAP3D/common/algorithms/CurveTool.java

```
@@ -15,130 +15,78 @@
-				writer.write(buffer.toString());
-				buffer = new StringBuffer();
-			}
-			writer.close();
-		}
-		
-		catch(Exception e){
-			
-		}
-		
-	}
+
+  private SliceSparseMatrix objSliceSparseMatrix;
+
+  /**
+   * <p>
+   * CurveTool constructor takes object of SliceSparseMatrix class.
+   * </p>
+   */
+  public CurveTool(SliceSparseMatrix sliceSparseMatrix) {
+    objSliceSparseMatrix = sliceSparseMatrix;
+  }
+
+  /**
+   * <p>
+   * estimateFwhmMs method estimates the FWHM for given number of random scans.
+   * 
+   * @param numberOfScansForFWHMCalc a {@link java.lang.Integer} object. This is number scans
+   *        required for estimation of fwhm.
+   * 
+   * @return fwhm a {@link java.lang.Double} object.This is Full width half maximum.
+   *         </p>
+   */
+  public double estimateFwhmMs(int numberOfScansForFWHMCalc) {
+
+    double sigma = 0;
+    int countProperIteration = 0;
+    int countTotalIteration = 0;
+
+    while (countProperIteration < numberOfScansForFWHMCalc) {
+      countTotalIteration++;
+
+
+      if (countTotalIteration > objSliceSparseMatrix.getSizeOfRawDataFile()) {
+        System.out.println(countTotalIteration);
+        throw new IllegalArgumentException("Cannot calculate FWHM.");
+      }
+
+      Random generator = new Random();
+      int randInt = generator.nextInt(objSliceSparseMatrix.getSizeOfRawDataFile());
+      List<SliceSparseMatrix.VerticalSliceDataPoint> verticalSlice =
+          objSliceSparseMatrix.getVerticalSlice(randInt);
+
+      if (verticalSlice == null)
+        continue;
+
+      WeightedObservedPoints obs = new WeightedObservedPoints();
+
+      for (SliceSparseMatrix.VerticalSliceDataPoint datapoint : verticalSlice) {
+        obs.add(datapoint.mz, datapoint.intensity);
+      }
+
+      try {
+        double[] parameters = GaussianCurveFitter.create().fit(obs.toList());
+        sigma += 2.35482 * parameters[2];
+
+      } catch (MathIllegalArgumentException e) {
+        continue;
```

As @tomas-pluskal said before, don't just swallow this exception. If this exception is thrown, is it fatal and you should throw a custom ```MSDKException``` or can you add a default value to ```sigma```?